### ISSUE TITLE
Don't announce small double turns

### ISSUE THREAD

**Issue by scaidermern** [osmandapp/OsmAnd#15195]
Sometimes you have to turn left and immediately right again (or vice versa), for example in the following situation:

![foo](https://user-images.githubusercontent.com/1040784/187635346-80812572-21a7-47e4-a8db-a8f099d7c13b.jpg)
https://www.openstreetmap.org/#map=18/51.31641/13.11360

OsmAnd currently announces this as something like "turn left, then keep right". This can be quite confusing for the user since in reality these intersections often appear very close to a straight line. Instead, OsmAnd could just omit such small double turns and simply announce "go straight ahead". This will be both shorter and clearer in many cases.

**Comment by vshcherb** [comment]
Very tricky... Potentially we could have bugs due to non-precise micromapping.

**Comment by scaidermern** [comment]
I agree that some cases might be tricky. However, overall I think skipping such announcements (up to a certain distance between double turns) might have more advantages than disadvantages.

I'd like to hear other opinions about this.

**Comment by mikehgentry** [comment]
Some wording like "Cross X Street and continue along Y street" might be even clearer? Obviously only if it's actually a different road.

**Comment by dktue** [comment]
I agree that detection of those double turns is hard but it's definitely something I also did observe myself. Would be a welcome improvement.

**Comment by sonora** [comment]
We do something similar already by combining successive turns towards the same direction if within a certain threshold distance., i.e. that method could be tweaked. Trouble is it is hard to judge what ground truth looks like, I guess there may be cases where two separate command are actually helpful...

**Comment by sre** [comment]
These small double turns happen regularly for me in the bicycle routing profile when the route changes from a bicycle way mapped separately onto the street itself (e.g. because the bicycle way becomes a pedestrian way). It is quite confusing when one cannot see the screen and easily leads to taken wrong turns.

**Comment by vmicho** [comment]
> These small double turns happen regularly for me in the bicycle routing profile when the route changes from a bicycle way mapped separately onto the street itself (e.g. because the bicycle way becomes a pedestrian way). It is quite confusing when one cannot see the screen and easily leads to taken wrong turns.

this should be a separate issue. I have also problems with joining road or taking outgoing cyclepath. I am thinking about this for some time already ...

**Comment by vmicho** [comment]
"turn left, then keep right"
That is announced also when the next right turn comes in like 50 m. Very confusing (and in case of the next turn is like 20+m quite useless).
If the next opposite turn is in like 20 m or less the announcement should differ to somwhow announce that zig-zag maneuver (e.g. "turn left and after 5 m turn (slight/sharply)right to continue in the same direction|street x). If >=20m then it should be only a standard separate 1 turn announcement. (note: the value 20 m is my right-now uneducated guess)

**Comment by andreasscherbaum** [comment]
This issue also happens when crossing a regular street while walking. The general navigation is "go straight", but the navigation turns left and right several times on short occasion, instead of saying "cross the street and go straight".

Some examples from Tokyo, it happened in other cities as well:

<img width="692" height="500" alt="Image" src="https://github.com/user-attachments/assets/835d5b8c-d06e-45ef-b42e-a66715e634bd" />

<img width="468" height="397" alt="Image" src="https://github.com/user-attachments/assets/26c3be86-da16-486d-80c3-441e3863c74a" />

<img width="557" height="341" alt="Image" src="https://github.com/user-attachments/assets/a020037f-33c7-4b72-a88a-1bfbdfca4b66" />

**Comment by anten1222** [comment]
The fact that the voice prompts don't match the directional arrows is a widespread and significant problem that annoys everyone. Will it ever be fixed?

**Comment by Peter0** [comment]
Maybe there should be some algorithm that detects a structure on the map that one would call a junction.
Then give an announcement that expresses somehow an angle in which to leave the junction. 
There should be several modes how to announce the angle. My uneducated first guess as my favorite would be to give the announcement in a "o'clock manner". Like "go 12 o'clock", which means straight or "9 o'clock", which means turn left. And then give an announcement like "Leave the junction ahead at 9 o'clock direction".
Another mode, more for the compass lovers, could be an absolute bearing. Like "leave the junction on a course bearing of 275°".
And I think another favorite would be a combined mode like "straight", "on the left", "on the right" for simple cases of rectangular crossing and the o'clock-directions for more complicated cases of non-rectangular or more than 4 streets junction or something like that. (Note that "leave on the left" doesn't necessarily means "turn left", e.g. at a roundabout (@ right-hand traffic).)

Just my two cents.

**Comment by scaidermern** [comment]
> Maybe there should be some algorithm that detects a structure on the map that one would call a junction.
> Then give an announcement that expresses somehow an angle in which to leave the junction.

This sounds much more useful than the current approach. For blind people it totally makes sense to announce every little turn. For other people it doesn't. Instead, providing the general direction for passing a junction is much more useful and less irritating. If I have to pass a junction straight then please tell me just exact this.

**Comment by Peter0** [comment]
> For blind people …

I am not a blind one, but in the area I am acquainted with, the mapping is, coarsely speaking, not precise enough to justify the promises made by this granular directions.

**Comment by anten1222** [comment]
I am a frequent cyclist.

The instruction

"Keep right in meters and then left" is not only unnecessary but also dangerous and should be removed.
